usr.sbin/config/config.5
.\" Copyright (c) 2003 Joseph Koshy
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 17, 2022
.Dt CONFIG 5
.Os
.Sh NAME
.Nm config
.Nd kernel configuration file format
.Sh DESCRIPTION
A kernel configuration file specifies the configuration of a
.Fx
kernel.
It is processed by
.Xr config 8
to create a build environment where a kernel may be built using
.Xr make 1 .
.Ss Lexical Structure
A kernel configuration file comprises a sequence of specification
directives.
.Pp
A specification directive starts with a keyword at the beginning
of the line and is followed by additional parameters.
.Pp
A specification directive may be terminated by a semicolon
.Ql \&;
or by a newline.
Long input lines may be broken into shorter lines by starting the
second and subsequent lines with a white space character.
.Pp
Case is significant,
.Dq Li machine
and
.Dq Li MACHINE
are different tokens.
.Pp
A double quote character
.Ql \[dq]
starts a quoted string.
All characters up to the next quote character form the value
of the quoted string.
A
.Ql \[dq]
character may be inserted into a quoted string by
using the sequence
.Ql \e\[dq] .
.Pp
Numbers are specified using
.Tn C Ns -style
syntax.
.Pp
A
.Ql #
character starts a comment; all characters from the
.Ql #
character till the end of the current line are ignored.
.Pp
Whitespace between tokens is ignored, except inside quoted strings.
Whitespace following a comment line is ignored.
.Ss Configuration Directives
Kernel configuration directives may appear in any order
in a kernel configuration file.
Directives are processed in order of appearance with subsequent
directive lines overriding the effect of prior ones.
.Pp
The list of keywords and their meanings are as follows:
.Pp
.Bl -tag -width indent -compact
.\" -------- CPU --------
.It Ic cpu Ar cputype
Specify the CPU this kernel will run on.
There can be more than one
.Ic cpu
directive in a configuration file.
The allowed list of CPU names is architecture specific and is
defined in the file
.Pa sys/conf/options . Ns Aq Ar arch .
.\" -------- DEVICE --------
.Pp
.It Ic device Ar name Op , Ar name Op ...
.It Ic devices Ar name Op , Ar name Op ...
Configures the specified devices
for inclusion into the kernel image.
Devices that are common to all architectures are
defined in the file
.Pa sys/conf/files .
Devices that are specific to architecture
.Ar arch
are defined in the file
.Pa sys/conf/files . Ns Aq Ar arch .
.\" -------- ENV --------
.Pp
.It Ic env Ar filename
Specifies a filename containing a kernel environment definition.
.Pp
The kernel will augment this compiled-in environment with the environment
prepared for it at boot time by
.Xr loader 8 .
Environment variables specified in the
.Xr loader 8
environment will take precedence over environment variables specified in
.Ar filename ,
and environment variables specified in the dynamic environment take precedence
over both of these.
.Pp
.Va loader_env.disabled=1
may be specified in the static environment to disable the
.Xr loader 8
environment.
Disabling the
.Xr loader 8
environment should be done with caution and due consideration for whether or not
it supplies environment variables needed for properly booting the system.
.Pp
.Va static_env.disabled=1
may be specified in the
.Xr loader 8
environment to disable use of the static environment.
This option has no effect if specified in any environment after the
.Xr loader 8
environment is processed.
This option is not usable in conjunction with
.Va loader_env.disabled .
.Pp
This directive is useful for setting kernel tunables in
embedded environments that do not start from
.Xr loader 8 .
.Pp
All
.Ic env
and
.Ic envvar
directives will be processed and added to the static environment in reversed
order of appearance so that later specified variables properly override earlier
specified variables.
Note that within
.Ar filename ,
the first appearance of a given variable will be the first one seen by the
kernel, effectively shadowing any later appearances of the same variable within
.Ar filename .
.\" -------- ENVVAR --------
.Pp
.It Ic envvar Ar setting
Specifies an individual environment setting to be added to the kernel's
compiled-in environment.
.Ar setting
must be of the form
.Dq Va name=value .
Optional quotes are supported in both name and value.
.Pp
All
.Ic env
and
.Ic envvar
directives will be processed and added to the static environment in reversed
order of appearance so that later specified variables properly override earlier
specified variables.
.\" -------- FILES --------
.Pp
.It Ic files Ar filename
Specifies a file containing a list of files specific to that kernel
configuration file (a la
.Pa files . Ns Aq Ar arch ) .
.\" -------- HINTS --------
.Pp
.It Ic hints Ar filename
Specifies a file to load a static device configuration specification
from.
From
.Fx 5.0
onwards, the kernel reads the system's device configuration at boot
time (see
.Xr device.hints 5 ) .
This directive configures the kernel to use the static device configuration
listed in
.Ar filename .
.Pp
Hints provided in this static device configuration will be overwritten in the
order in which they're encountered.
Hints in the compiled-in environment takes precedence over compiled-in hints,
and hints in the environment prepared for the kernel by
.Xr loader 8
takes precedence over hints in the compiled-in environment.
.Pp
Once the dynamic environment becomes available, all compiled-in hints will be
added to the dynamic environment if they do not already have an override in
the dynamic environment.
The dynamic environment will then be used for all searches of hints.
.Pp
.Va loader_hints.disabled=1
may be specified in the compiled-in environment to disable use of hints from
the
.Xr loader 8
environment.
Disabling hints from the
.Xr loader 8
environment should be done with caution and due consideration for whether or not
it supplies hints needed for properly booting the system.
.Pp
.Va static_hints.disabled=1
may be specified in either a compiled-in environment or the
.Xr loader 8
environment to disable use of these hints files.
This option has no effect if specified in any environment after the
.Xr loader 8
environment is processed.
This option also has no effect if loader hints have been disabled by the static
environment setting
.Va loader_hints.disabled=1 .
.Pp
The file
.Ar filename
must conform to the syntax specified by
.Xr device.hints 5 .
Multiple hints lines are allowed.
The resulting hints will be the files concatenated in reverse order of
appearance so that hints in later files properly override hints in earlier
files.
.\" -------- IDENT --------
.Pp
.It Ic ident Ar name
Set the kernel name to
.Ar name .
At least one
.Ic ident
directive is required.
.\" -------- INCLUDE --------
.Pp
.It Ic include Ar filename
Read subsequent text from file
.Ar filename
and return to the current file after
.Ar filename
is successfully processed.
.\" -------- MACHINE --------
.Pp
.It Ic machine Ar arch Op Ar cpuarch
Specifies the architecture of the machine the kernel is being
compiled for.
Legal values for
.Ar arch
include:
.Pp
.Bl -tag -width ".Cm powerpc" -compact
.It Cm arm64
The 64-bit ARM application architecture.
.It Cm arm
The ARM architecture
.It Cm amd64
The AMD x86-64 architecture.
.It Cm i386
The Intel x86 based PC architecture.
.It Cm powerpc
The IBM PowerPC architecture.
.It Cm riscv
The RISC-V architecture.
.El
.Pp
If argument
.Ar cpuarch
is specified, it points
.Xr config 8
to the cpu architecture of the machine.
When
.Ar cpuarch
is not specified, it is assumed to be the same as
.Ar arch .
.Ar arch
corresponds to MACHINE.
.Ar cpuarch
corresponds to MACHINE_ARCH.
.Pp
A kernel configuration file may have only one
.Ic machine
directive, unless the second one matches the
machine argument in the first one exactly.
.\" -------- MAKEOPTION --------
.Pp
.It Ic makeoption Ar options
.It Ic makeoptions Ar options
Add
.Ar options
to the generated makefile.
.Pp
The
.Ar options
argument is a comma separated list of one or more option
specifications.
Each option specification has the form
.Pp
.D1 Ar MakeVariableName Ns Op = Ns Ar Value
.D1 Ar MakeVariableName Ns += Ns Ar Value
.Pp
and results in the appropriate
.Xr make 1
variable definition being inserted into the generated makefile.
If only the name of the
.Xr make 1
variable is specified,
.Ar value
is assumed to be the empty string.
.Pp
Example:
.Bd -literal -offset indent -compact
makeoptions MYMAKEOPTION="foo"
makeoptions MYMAKEOPTION+="bar"
makeoptions MYNULLMAKEOPTION
.Ed
.\" -------- MAXUSERS --------
.Pp
.It Ic maxusers Ar number
This optional directive is used to configure the size
of some kernel data structures.
The parameter
.Ar number
can be 0 (the default) or an integer greater than or equal to 2.
A value of 0 indicates that the kernel should configure
its data structures according to the size of available
physical memory.
If auto configuration is requested, the kernel will set
this tunable to a value between 32 and 384 for 32-bit systems,
or scale the value higher based on available memory for 64-bit
systems.
.Pp
As explained in
.Xr tuning 7 ,
this tunable can also be set at boot time using
.Xr loader 8 .
.\" -------- NOCPU --------
.Pp
.It Ic nocpu Ar cputype
Remove the specified CPU
from the list of previously selected CPUs.
This directive can be used to cancel the effect of
.Ic cpu
directives in files included using
.Ic include .
.\" -------- NODEVICE --------
.Pp
.It Ic nodevice Ar name Op , Ar name Op ...
.It Ic nodevices Ar name Op , Ar name Op ...
Remove the specified devices
from the list of previously selected devices.
This directive can be used to cancel the effects of
.Ic device
or
.Ic devices
directives in files included using
.Ic include .
.\" -------- NOMAKEOPTION --------
.Pp
.It Ic nomakeoption Ar name
.It Ic nomakeoptions Ar name
Removes previously defined
.Xr make 1
option
.Ar name
from the kernel build.
This directive can be used to cancel the effects of
.Ic makeoption
directives in files included using
.Ic include .
.\" -------- NOOPTION --------
.Pp
.It Ic nooption Ar name Op , Ar name Op ...
.It Ic nooptions Ar name Op , Ar name Op ...
Remove the specified kernel options
from the list of previously defined options.
This directive can be used to cancel the effects of
.Ic option
or
.Ic options
directives in files included using
.Ic include .
.\" -------- OPTIONS --------
.Pp
.It Ic option Ar optionspec Op , Ar optionspec Op ...
.It Ic options Ar optionspec Op , Ar optionspec Op ...
Add compile time kernel options to the kernel build.
Each option specification has the form
.Pp
.D1 Ar name Ns Op = Ns Ar value
.Pp
If
.Ar value
is not specified, it is assumed to be
.Dv NULL .
Options common to all architectures are specified in
the file
.Pa sys/conf/options .
Options specific to architecture
.Ar arch
are specified in the file
.Pa sys/conf/options . Ns Aq Ar arch .
.El
.Sh FILES
.Bl -tag -width ".Pa sys/conf/Makefile. Ns Ar arch" -compact
.It Pa sys/compile/ Ns Ar NAME
Compile directory created from a kernel configuration.
.It Pa sys/conf/Makefile . Ns Ar arch
.Pa Makefile
fragments for architecture
.Ar arch .
.It Pa sys/conf/files
Devices common to all architectures.
.It Pa sys/conf/files . Ns Ar arch
Devices for architecture
.Ar arch .
.It Pa sys/conf/options
Options common to all architectures.
.It Pa sys/conf/options . Ns Ar arch
Options for architecture
.Ar arch .
.El
.Sh SEE ALSO
.Xr kenv 1 ,
.Xr make 1 ,
.Xr device.hints 5 ,
.Xr loader.conf 5 ,
.Xr config 8 ,
.Xr kldload 8 ,
.Xr loader 8
.Rs
.%T "Building 4.4BSD Kernels with Config"
.%A "Samuel J. Leffler"
.%A "Michael J. Karels"
.Re
.Sh HISTORY
The
.Xr config 8
utility first appeared in
.Bx 4.1 ,
and was subsequently revised in
.Bx 4.4 .
.Pp
The kernel configuration mechanism changed further in
.Fx 4.0
and
.Fx 5.0 ,
moving toward an architecture supporting dynamic kernel
configuration.


stand/lua/config.lua
--
-- SPDX-License-Identifier: BSD-2-Clause-FreeBSD
--
-- Copyright (c) 2015 Pedro Souza <pedrosouza@freebsd.org>
-- Copyright (c) 2018 Kyle Evans <kevans@FreeBSD.org>
-- All rights reserved.
--
-- Redistribution and use in source and binary forms, with or without
-- modification, are permitted provided that the following conditions
-- are met:
-- 1. Redistributions of source code must retain the above copyright
--    notice, this list of conditions and the following disclaimer.
-- 2. Redistributions in binary form must reproduce the above copyright
--    notice, this list of conditions and the following disclaimer in the
--    documentation and/or other materials provided with the distribution.
--
-- THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
-- ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
-- IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
-- ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
-- FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
-- DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
-- OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
-- HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
-- LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
-- OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
-- SUCH DAMAGE.
--
-- $FreeBSD$
--

local hook = require("hook")

local config = {}
local modules = {}
local carousel_choices = {}
-- Which variables we changed
local env_changed = {}
-- Values to restore env to (nil to unset)
local env_restore = {}

local MSG_FAILEXEC = "Failed to exec '%s'"
local MSG_FAILSETENV = "Failed to '%s' with value: %s"
local MSG_FAILOPENCFG = "Failed to open config: '%s'"
local MSG_FAILREADCFG = "Failed to read config: '%s'"
local MSG_FAILPARSECFG = "Failed to parse config: '%s'"
local MSG_FAILEXBEF = "Failed to execute '%s' before loading '%s'"
local MSG_FAILEXAF = "Failed to execute '%s' after loading '%s'"
local MSG_MALFORMED = "Malformed line (%d):\n\t'%s'"
local MSG_DEFAULTKERNFAIL = "No kernel set, failed to load from module_path"
local MSG_KERNFAIL = "Failed to load kernel '%s'"
local MSG_XENKERNFAIL = "Failed to load Xen kernel '%s'"
local MSG_XENKERNLOADING = "Loading Xen kernel..."
local MSG_KERNLOADING = "Loading kernel..."
local MSG_MODLOADING = "Loading configured modules..."
local MSG_MODBLACKLIST = "Not loading blacklisted module '%s'"

local MODULEEXPR = '([%w-_]+)'
local QVALEXPR = "\"([%w%s%p]-)\""
local QVALREPL = QVALEXPR:gsub('%%', '%%%%')
local WORDEXPR = "([%w]+)"
local WORDREPL = WORDEXPR:gsub('%%', '%%%%')

local function restoreEnv()
	-- Examine changed environment variables
	for k, v in pairs(env_changed) do
		local restore_value = env_restore[k]
		if restore_value == nil then
			-- This one doesn't need restored for some reason
			goto continue
		end
		local current_value = loader.getenv(k)
		if current_value ~= v then
			-- This was overwritten by some action taken on the menu
			-- most likely; we'll leave it be.
			goto continue
		end
		restore_value = restore_value.value
		if restore_value ~= nil then
			loader.setenv(k, restore_value)
		else
			loader.unsetenv(k)
		end
		::continue::
	end

	env_changed = {}
	env_restore = {}
end

local function setEnv(key, value)
	-- Track the original value for this if we haven't already
	if env_restore[key] == nil then
		env_restore[key] = {value = loader.getenv(key)}
	end

	env_changed[key] = value

	return loader.setenv(key, value)
end

-- name here is one of 'name', 'type', flags', 'before', 'after', or 'error.'
-- These are set from lines in loader.conf(5): ${key}_${name}="${value}" where
-- ${key} is a module name.
local function setKey(key, name, value)
	if modules[key] == nil then
		modules[key] = {}
	end
	modules[key][name] = value
end

-- Escapes the named value for use as a literal in a replacement pattern.
-- e.g. dhcp.host-name gets turned into dhcp%.host%-name to remove the special
-- meaning.
local function escapeName(name)
	return name:gsub("([%p])", "%%%1")
end

local function processEnvVar(value)
	for name in value:gmatch("${([^}]+)}") do
		local replacement = loader.getenv(name) or ""
		value = value:gsub("${" .. escapeName(name) .. "}", replacement)
	end
	for name in value:gmatch("$([%w%p]+)%s*") do
		local replacement = loader.getenv(name) or ""
		value = value:gsub("$" .. escapeName(name), replacement)
	end
	return value
end

local function checkPattern(line, pattern)
	local function _realCheck(_line, _pattern)
		return _line:match(_pattern)
	end

	if pattern:find('$VALUE') then
		local k, v, c
		k, v, c = _realCheck(line, pattern:gsub('$VALUE', QVALREPL))
		if k ~= nil then
			return k,v, c
		end
		return _realCheck(line, pattern:gsub('$VALUE', WORDREPL))
	else
		return _realCheck(line, pattern)
	end
end

-- str in this table is a regex pattern.  It will automatically be anchored to
-- the beginning of a line and any preceding whitespace will be skipped.  The
-- pattern should have no more than two captures patterns, which correspond to
-- the two parameters (usually 'key' and 'value') that are passed to the
-- process function.  All trailing characters will be validated.  Any $VALUE
-- token included in a pattern will be tried first with a quoted value capture
-- group, then a single-word value capture group.  This is our kludge for Lua
-- regex not supporting branching.
--
-- We have two special entries in this table: the first is the first entry,
-- a full-line comment.  The second is for 'exec' handling.  Both have a single
-- capture group, but the difference is that the full-line comment pattern will
-- match the entire line.  This does not run afoul of the later end of line
-- validation that we'll do after a match.  However, the 'exec' pattern will.
-- We document the exceptions with a special 'groups' index that indicates
-- the number of capture groups, if not two.  We'll use this later to do
-- validation on the proper entry.
--
local pattern_table = {
	{
		luaexempt = true,
		str = "(#.*)",
		process = function(_, _)  end,
		groups = 1,
	},
	--  module_load="value"
	{
		name = MODULEEXPR .. "_load%s*",
		val = "%s*$VALUE",
		process = function(k, v)
			if modules[k] == nil then
				modules[k] = {}
			end
			modules[k].load = v:upper()
		end,
	},
	--  module_name="value"
	{
		name = MODULEEXPR .. "_name%s*",
		val = "%s*$VALUE",
		process = function(k, v)
			setKey(k, "name", v)
		end,
	},
	--  module_type="value"
	{
		name = MODULEEXPR .. "_type%s*",
		val = "%s*$VALUE",
		process = function(k, v)
			setKey(k, "type", v)
		end,
	},
	--  module_flags="value"
	{
		name = MODULEEXPR .. "_flags%s*",
		val = "%s*$VALUE",
		process = function(k, v)
			setKey(k, "flags", v)
		end,
	},
	--  module_before="value"
	{
		name = MODULEEXPR .. "_before%s*",
		val = "%s*$VALUE",
		process = function(k, v)
			setKey(k, "before", v)
		end,
	},
	--  module_after="value"
	{
		name = MODULEEXPR .. "_after%s*",
		val = "%s*$VALUE",
		process = function(k, v)
			setKey(k, "after", v)
		end,
	},
	--  module_error="value"
	{
		name = MODULEEXPR .. "_error%s*",
		val = "%s*$VALUE",
		process = function(k, v)
			setKey(k, "error", v)
		end,
	},
	--  exec="command"
	{
		luaexempt = true,
		name = "exec%s*",
		val = "%s*" .. QVALEXPR,
		process = function(k, _)
			if cli_execute_unparsed(k) ~= 0 then
				print(MSG_FAILEXEC:format(k))
			end
		end,
		groups = 1,
	},
	--  env_var="value"
	{
		name = "([%w%p]+)%s*",
		val = "%s*$VALUE",
		process = function(k, v)
			if setEnv(k, processEnvVar(v)) ~= 0 then
				print(MSG_FAILSETENV:format(k, v))
			end
		end,
	},
	--  env_var=num
	{
		name = "([%w%p]+)%s*",
		val = "%s*(-?%d+)",
		process = function(k, v)
			if setEnv(k, processEnvVar(v)) ~= 0 then
				print(MSG_FAILSETENV:format(k, tostring(v)))
			end
		end,
	},
}

local function isValidComment(line)
	if line ~= nil then
		local s = line:match("^%s*#.*")
		if s == nil then
			s = line:match("^%s*$")
		end
		if s == nil then
			return false
		end
	end
	return true
end

local function getBlacklist()
	local blacklist = {}
	local blacklist_str = loader.getenv('module_blacklist')
	if blacklist_str == nil then
		return blacklist
	end

	for mod in blacklist_str:gmatch("[;, ]?([%w-_]+)[;, ]?") do
		blacklist[mod] = true
	end
	return blacklist
end

local function loadModule(mod, silent)
	local status = true
	local blacklist = getBlacklist()
	local pstatus
	for k, v in pairs(mod) do
		if v.load ~= nil and v.load:lower() == "yes" then
			local module_name = v.name or k
			if blacklist[module_name] ~= nil then
				if not silent then
					print(MSG_MODBLACKLIST:format(module_name))
				end
				goto continue
			end
			if not silent then
				loader.printc(module_name .. "...")
			end
			local str = "load "
			if v.type ~= nil then
				str = str .. "-t " .. v.type .. " "
			end
			str = str .. module_name
			if v.flags ~= nil then
				str = str .. " " .. v.flags
			end
			if v.before ~= nil then
				pstatus = cli_execute_unparsed(v.before) == 0
				if not pstatus and not silent then
					print(MSG_FAILEXBEF:format(v.before, k))
				end
				status = status and pstatus
			end

			if cli_execute_unparsed(str) ~= 0 then
				-- XXX Temporary shim: don't break the boot if
				-- loader hadn't been recompiled with this
				-- function exposed.
				if loader.command_error then
					print(loader.command_error())
				end
				if not silent then
					print("failed!")
				end
				if v.error ~= nil then
					cli_execute_unparsed(v.error)
				end
				status = false
			elseif v.after ~= nil then
				pstatus = cli_execute_unparsed(v.after) == 0
				if not pstatus and not silent then
					print(MSG_FAILEXAF:format(v.after, k))
				end
				if not silent then
					print("ok")
				end
				status = status and pstatus
			end
		end
		::continue::
	end

	return status
end

local function readFile(name, silent)
	local f = io.open(name)
	if f == nil then
		if not silent then
			print(MSG_FAILOPENCFG:format(name))
		end
		return nil
	end

	local text, _ = io.read(f)
	-- We might have read in the whole file, this won't be needed any more.
	io.close(f)

	if text == nil and not silent then
		print(MSG_FAILREADCFG:format(name))
	end
	return text
end

local function checkNextboot()
	local nextboot_file = loader.getenv("nextboot_conf")
	if nextboot_file == nil then
		return
	end

	local text = readFile(nextboot_file, true)
	if text == nil then
		return
	end

	if text:match("^nextboot_enable=\"NO\"") ~= nil then
		-- We're done; nextboot is not enabled
		return
	end

	if not config.parse(text) then
		print(MSG_FAILPARSECFG:format(nextboot_file))
	end

	-- Attempt to rewrite the first line and only the first line of the
	-- nextboot_file. We overwrite it with nextboot_enable="NO", then
	-- check for that on load.
	-- It's worth noting that this won't work on every filesystem, so we
	-- won't do anything notable if we have any errors in this process.
	local nfile = io.open(nextboot_file, 'w')
	if nfile ~= nil then
		-- We need the trailing space here to account for the extra
		-- character taken up by the string nextboot_enable="YES"
		-- Or new end quotation mark lands on the S, and we want to
		-- rewrite the entirety of the first line.
		io.write(nfile, "nextboot_enable=\"NO\" ")
		io.close(nfile)
	end
end

local function processEnv(k, v)
	-- Perhaps we should split out the name portion of the expressions in
	-- the pattern table, but for now this is OK.
	for _, val in ipairs(pattern_table) do
		if val.luaexempt then
			goto next
		end

		local matched = k:match(assert(val.name))

		if matched then
			val.process(matched, v)
		end
		::next::
	end
end

-- Module exports
config.verbose = false

-- The first item in every carousel is always the default item.
function config.getCarouselIndex(id)
	return carousel_choices[id] or 1
end

function config.setCarouselIndex(id, idx)
	carousel_choices[id] = idx
end

-- Returns true if we processed the file successfully, false if we did not.
-- If 'silent' is true, being unable to read the file is not considered a
-- failure.
function config.processFile(name, silent)
	if silent == nil then
		silent = false
	end

	local text = readFile(name, silent)
	if text == nil then
		return silent
	end

	if name:match(".lua$") then
		local cfg_env = {}

		-- Give local modules a chance to populate the config
		-- environment.
		hook.runAll("config.buildenv", cfg_env)
		function cfg_env.getenv(k)
			if cfg_env[k] ~= nil then
				return cfg_env[k]
			end
			return loader.getenv(k)
		end
		function cfg_env.setenv(k, v)
			if type(k) == "table" then
				if v ~= nil then
					error(
					    "setenv: illegal version, second arg not valid with a table")
				end
				for tk, tv in pairs(k) do
					cfg_env[tk] = tv
				end
			else
				cfg_env[k] = v
			end
		end
		local res, err = pcall(load(text, name, "t", cfg_env))
		if res then
			for k, v in pairs(cfg_env) do
				if type(v) ~= "function" then
					processEnv(k, v)
				end
			end
		else
			print(err)
		end

		return res
	else
		return config.parse(text)
	end
end

-- silent runs will not return false if we fail to open the file
function config.parse(text)
	local n = 1
	local status = true

	for line in text:gmatch("([^\n]+)") do
		if line:match("^%s*$") == nil then
			for _, val in ipairs(pattern_table) do
				if val.str == nil then
					val.str = val.name .. "=" .. val.val
				end
				local pattern = '^%s*' .. val.str .. '%s*(.*)';
				local cgroups = val.groups or 2
				local k, v, c = checkPattern(line, pattern)
				if k ~= nil then
					-- Offset by one, drats
					if cgroups == 1 then
						c = v
						v = nil
					end

					if isValidComment(c) then
						val.process(k, v)
						goto nextline
					end

					break
				end
			end

			print(MSG_MALFORMED:format(n, line))
			status = false
		end
		::nextline::
		n = n + 1
	end

	return status
end

function config.readConfFiles(files, loaded_files)
	if files ~= nil then
		-- The caller may not have passed in loader_conf_files; we could
		-- have instead gotten some other string of files.  We don't
		-- want to trigger any redundant re-read/loads based on this.
		local prefiles = loader.getenv("loader_conf_files")
		for name in files:gmatch("([%w%p]+)%s*") do
			if loaded_files[name] ~= nil then
				goto continue
			end

			print("Loading " .. name)
			-- These may or may not exist, and that's ok. Do a
			-- silent parse so that we complain on parse errors but
			-- not for them simply not existing.
			if not config.processFile(name, true) then
				print(MSG_FAILPARSECFG:format(name))
			end

			loaded_files[name] = true
			local newfiles = loader.getenv("loader_conf_files")
			if prefiles ~= newfiles then
				-- Recurse; process the new files immediately.
				-- If we come back and it turns out we've
				-- already loaded the rest of what was in the
				-- original loader_conf_files, no big deal.
				config.readConfFiles(newfiles, loaded_files)
				prefiles = newfiles
			end
			::continue::
		end
	end
end

-- other_kernel is optionally the name of a kernel to load, if not the default
-- or autoloaded default from the module_path
function config.loadKernel(other_kernel)
	local flags = loader.getenv("kernel_options") or ""
	local kernel = other_kernel or loader.getenv("kernel")

	local function tryLoad(names)
		for name in names:gmatch("([^;]+)%s*;?") do
			local r = loader.perform("load " .. name ..
			     " " .. flags)
			if r == 0 then
				return name
			end
		end
		return nil
	end

	local function getModulePath()
		local module_path = loader.getenv("module_path")
		local kernel_path = loader.getenv("kernel_path")

		if kernel_path == nil then
			return module_path
		end

		-- Strip the loaded kernel path from module_path. This currently assumes
		-- that the kernel path will be prepended to the module_path when it's
		-- found.
		kernel_path = escapeName(kernel_path .. ';')
		return module_path:gsub(kernel_path, '')
	end

	local function loadBootfile()
		local bootfile = loader.getenv("bootfile")

		-- append default kernel name
		if bootfile == nil then
			bootfile = "kernel"
		else
			bootfile = bootfile .. ";kernel"
		end

		return tryLoad(bootfile)
	end

	-- kernel not set, try load from default module_path
	if kernel == nil then
		local res = loadBootfile()

		if res ~= nil then
			-- Default kernel is loaded
			config.kernel_loaded = nil
			return true
		else
			print(MSG_DEFAULTKERNFAIL)
			return false
		end
	else
		-- Use our cached module_path, so we don't end up with multiple
		-- automatically added kernel paths to our final module_path
		local module_path = getModulePath()
		local res

		if other_kernel ~= nil then
			kernel = other_kernel
		end
		-- first try load kernel with module_path = /boot/${kernel}
		-- then try load with module_path=${kernel}
		local paths = {"/boot/" .. kernel, kernel}

		for _, v in pairs(paths) do
			loader.setenv("module_path", v)
			res = loadBootfile()

			-- succeeded, add path to module_path
			if res ~= nil then
				config.kernel_loaded = kernel
				if module_path ~= nil then
					loader.setenv("module_path", v .. ";" ..
					    module_path)
					loader.setenv("kernel_path", v)
				end
				return true
			end
		end

		-- failed to load with ${kernel} as a directory
		-- try as a file
		res = tryLoad(kernel)
		if res ~= nil then
			config.kernel_loaded = kernel
			return true
		else
			print(MSG_KERNFAIL:format(kernel))
			return false
		end
	end
end

function config.selectKernel(kernel)
	config.kernel_selected = kernel
end

function config.load(file, reloading)
	if not file then
		file = "/boot/defaults/loader.conf"
	end

	if not config.processFile(file) then
		print(MSG_FAILPARSECFG:format(file))
	end

	local loaded_files = {file = true}
	config.readConfFiles(loader.getenv("loader_conf_files"), loaded_files)

	checkNextboot()

	local verbose = loader.getenv("verbose_loading") or "no"
	config.verbose = verbose:lower() == "yes"
	if not reloading then
		hook.runAll("config.loaded")
	end
end

-- Reload configuration
function config.reload(file)
	modules = {}
	restoreEnv()
	config.load(file, true)
	hook.runAll("config.reloaded")
end

function config.loadelf()
	local xen_kernel = loader.getenv('xen_kernel')
	local kernel = config.kernel_selected or config.kernel_loaded
	local status

	if xen_kernel ~= nil then
		print(MSG_XENKERNLOADING)
		if cli_execute_unparsed('load ' .. xen_kernel) ~= 0 then
			print(MSG_XENKERNFAIL:format(xen_kernel))
			return false
		end
	end
	print(MSG_KERNLOADING)
	if not config.loadKernel(kernel) then
		return false
	end
	hook.runAll("kernel.loaded")

	print(MSG_MODLOADING)
	status = loadModule(modules, not config.verbose)
	hook.runAll("modules.loaded")
	return status
end

hook.registerType("config.buildenv")
hook.registerType("config.loaded")
hook.registerType("config.reloaded")
hook.registerType("kernel.loaded")
hook.registerType("modules.loaded")
return config
